Approval Workflow documentation (#7668)
nautobot/nautobot: nautobot/docs/user-guide/platform-functionality/approval-workflow.md

# Approval Workflows

Approval Workflows allows for a multi-stage review and approval of processes before making changes, running or creating specific objects in the system. They are defined in advance and attached to specific models based on certain constraints.

## Use cases

> - Preventing accidental deletion of critical data by requiring manager approval before deletion jobs run.

Author: On the other hand, how does such approval differ from what we already have described in `#### Approve or Deny a Stage`

### ApprovalWorkflowStageDefinition

A single approval stage within a workflow definition, specifying the approver group and required number of approvers.

### ApprovalWorkflow

A concrete instance of a workflow triggered for a specific object under review.

- Automatically changes to `Denied` if any stage is denied.
- Automatically changes to `Approved` if all stages are approved.
- Calls object’s `on_workflow_approved()` or `on_workflow_denied()` when final state is reached.

### ApprovalWorkflowStage

A stage instance within a workflow execution. Progresses from `Pending` → `Approved`/`Denied`.

- Approves automatically when minimum approvals are met.
- Denies immediately if any denial is submitted.
- Triggers parent workflow state updates.

### ApprovalWorkflowStageResponse

A single user's input on a specific stage - this may be an explicit decision (`approve`/`deny`) or simply a comment without a decision.

- Saving a response can trigger stage and workflow state updates.

## How-To Guides

### Attach Workflow to a Model Instance

> Workflows are automatically attached after creating, running, or updating an object that complies with the workflow model and constraints. Manual attachment is not available.

Author: Added here under `#### Constraints`

### Using Approval Workflow via UI

#### Create an Approval Workflow Definition with stages

1. Go to `Approvals > Workflow Definitions > Add Approval Workflow Definition`.
2. Enter:
    - **Name** (e.g., "Scheduled Job Run Workflow").
    - **Model** (e.g., `extras|scheduled job`).

>     - **Constraints** (optional JSON filter, e.g., `{"name": "Bulk Delete Objects Scheduled Job"}`).

Author: Could be, so maybe it's a good moment to implement it now in 3.0

    - **Current Stage** - Stage awaiting action.
    - **Actions Needed** - Remaining approvals required.
    - **State** - Current workflow state.
3. To approve the stage, select the green button.  
4. To deny the stage, select the red button.
5. After the decision action, a confirmation window appears where a comment can be added before confirming the action.

#### View My Requests

1. Open the **Approval Dashboard** and select the **My Requests** tab.
2. The table lists all workflows initiated by the current user. The columns include:
    - **Approval Workflow Definition** - Linked to the workflow definition.
    - **Object Type Under Review** - Model and object type for the request.
    - **Object Under Review** - Linked to the specific object awaiting or having received approval.
    - **Current State** - State of the workflow (e.g., `Pending`, `Approved`, `Denied`).
    - **User** - Requesting user.

#### Check Workflow State

1. Go to `Approvals > Approval Workflow Definition`
2. Select the required workflow definition.
3. In the **Workflows** list, all workflows for this definition are displayed.

> 4. Select the details button for a specific workflow.

Reviewer: So this is just for "interested third-parties" who aren't either a requester or an approver but want to see what's in flight for their own reasons?

http://nautobot/api/extras/approval-workflow-stages/?pending_my_approvals=true
```

- `?pending_my_approvals=false` — Returns stages the current user has already approved/denied.

```no-highlight
curl -X GET \
-H "Authorization: Token $TOKEN" \
-H "Accept: application/json; version=1.3; indent=4" \
http://nautobot/api/extras/approval-workflow-stages/?pending_my_approvals=false
```

If the parameter is omitted, all stages are returned regardless of approval status.

## Permissions by Persona

### Object Operator

Responsible for creating, updating, or running objects, which may trigger an approval workflow. Cannot approve workflows.

**Required permissions:**
- `extras.view_approvalworkflow` - to view created approval workflows after creating, updating, or running an object.
- `extras.view_approvalworkflowstage` - to view the **My Requests** tab on the Dashboard.
- `extras.view_approvalworkflowstageresponse` - to view responses for each stage.

### Approver

Responsible for reviewing and making decisions on approval workflow stages assigned to them. Must be a member of one or more approver groups (see [Approval Groups](#approver-groups)).

**Required permissions:**

> - `extras.view_approvalworkflowstage` and `extras.change_approvalworkflowstage` - to view the **My Approvals** tab on the Dashboard and take actions such as approve or deny.

Author: The disadvantage of this approach is that approver will see all stages and responses of a given flow. I am not sure if this is desirable.
<img width="1635" height="490" alt="Screenshot 2025-08-14 at 12 56 54" src="https://github.com/user-attachments/assets/d76675b7-cc1e-48d3-86dc-f02b71cbb538" />
user it's not a part of approver group in stage2, but can still see this stage if have the `view_approvalworkflow` permission

- Additional permissions:
    - `extras.view/add/change/delete_approvalworkflowdefinition`
    - `extras.view/add/change/delete_approvalworkflowstagedefinition`

### Approver Groups

Approval actions are controlled not only by permissions but also by **approver group membership** defined in each workflow definition:

- Each **Approval Workflow Stage Definition** specifies an **Approver Group** (user group authorized to act on that stage).
- A user must belong to the stage’s approver group to:
    - See the stage in the **My Approvals** tab.
    - Perform approval or denial actions.
- If the user has the required permissions but is **not** a member of the approver group, the stage will not appear in their dashboard and no actions will be available.

## Upgrade Considerations

**From Nautobot 2.x**  
  If upgrading from Nautobot 2.x, the management command `check_job_approval_status` is available to identify Jobs and Scheduled Jobs that have `approval_required=True`. Running this command prior to upgrading helps detect and address these cases by:
      - Approving (and running) or denying scheduled jobs that require approval.

>       - Defining approval workflows for Jobs where appropriate.

Reviewer 2: Can you validate this renders as you would expect on mkdocs?

    "extras.ApprovalWorkflowDefinition" }o--|| "contenttypes.ContentType": "applies to model"
    "extras.ApprovalWorkflowDefinition" ||--o{ "extras.ApprovalWorkflowStageDefinition": "defines stages"
    "extras.ApprovalWorkflowStageDefinition" }o--|| "users.Group": "approver group"
    "extras.ApprovalWorkflowDefinition" ||--o{ "extras.ApprovalWorkflow": "creates instances"
    "extras.ApprovalWorkflow" }o--|| "models.GenericForeignKey": "object under review"
    "extras.ApprovalWorkflow" ||--o{ "extras.ApprovalWorkflowStage": "has stages"
    "extras.ApprovalWorkflowStage" ||--o{ "extras.ApprovalWorkflowStageResponse": "has responses"
    "extras.ApprovalWorkflowStageResponse" }o--|| "users.User": "submitted by"
```

### ApprovalWorkflowDefinition

The template for a workflow, specifying which model(s) it applies to, any constraints, and the ordered list of stages. Represents a reusable definition of an approval workflow.

#### Constraints

Constraints define when a workflow definition should apply to a particular model instance.

- Constraints are stored as a dictionary of **field lookups → values** (e.g. `{"status": "active"}`).

> - Any valid Django ORM lookup can be used (e.g. `{"name__in": ["A", "B"]}`, `{"status__icontains": "active"}`, `{"priority__gte": 10}`), since constraints are passed directly into `.filter(**constraints)`.

Reviewer: `ScheduledJob` doesn't have a `status` or `priority` field, does it? Let's make sure the examples here are valid for the models that currently support workflows.

- Automatically changes to `Denied` if any stage is denied.
- Automatically changes to `Approved` if all stages are approved.
- Calls object’s `on_workflow_approved()` or `on_workflow_denied()` when final state is reached.

### ApprovalWorkflowStage

A concrete instance of stage within a workflow execution. Progresses from `Pending` → `Approved`/`Denied`.

- Approves automatically when minimum approvals are met.
- Denies immediately if any denial is submitted.
- Triggers parent workflow state updates.

### ApprovalWorkflowStageResponse

A single user's input on a specific stage - this may be an explicit decision (`approve`/`deny`) or simply a comment without a decision.

- Saving a response can trigger stage and workflow state updates.

## How-To Guides

### Attach Workflow to a Model Instance

Workflows are automatically attached after creating, running, or updating an object that complies with the workflow model and constraints. Manual attachment is not available.

### Using Approval Workflow via UI

#### Create an Approval Workflow Definition with stages

1. Go to `Approvals > Workflow Definitions > Add Approval Workflow Definition`.
    |   |   |
    |---|---|

>     | ![Navigation](../../media/development/approval-workflows/nav-approval-workflow-definition.png) | ![Workflow Definition](../../media/development/approval-workflows/approval-workflow-definition.png) |

Reviewer: This doesn't appear to be rendering correctly (https://docs-landing-page--7668.org.readthedocs.build/projects/core/en/7668/user-guide/platform-functionality/approval-workflow/#create-an-approval-workflow-definition-with-stages)

Author: fixed

    ![Add Approval Workflow Definition](../../media/development/approval-workflows/add-approval-workflow-definition.png)

2. Enter:

>     - **Name** (e.g., "Scheduled Job Run Workflow").

Reviewer 2: <img width="1125" height="171" alt="image" src="https://github.com/user-attachments/assets/3e5e0b80-d1b6-4147-8992-3a2f1c50fade" />

    - **Model** (e.g., `extras|scheduled job`).
    - **Constraints** (optional JSON filter, e.g., `{"job_model__name": "Bulk Delete Objects"}`, see [Contraints](#constraints)).
    - **Priority** (only one workflow definition (the highest, i.e. smallest priority one) will apply to any given object under review, see [Priority](#priority)).
3. In the **Approval Workflow Stage Definitions** section, define one or more stages:
    - **Stage Name** (a descriptive name for the approval step (e.g., "Finance Approval", "Security Review", "Management Approval")).

>     - **Weight** (order in which the stage is executed).

Reviewer: Conversely, I'd like to suggest that we change the workflow **stage** `weight` to something like `sequence` as it defines the _order in which the stages take effect_, not _which stage is more important than others_. @itdependsnetworks thoughts?

Author: I created a [ticket](https://networktocode.atlassian.net/browse/NAUTOBOT-1029) to discuss about it in next week when @itdependsnetworks will be back from PTO.

    - **Minimum Approvers** (number of approvals required).
    - **Approver Group** (group of users eligible to approve)
    - **Denial Message** (optional message shown if denied).

#### Approve or Deny a Stage

1. Go to `Approvals > Approval Dashboard` or on the right `User dropdown menu > Approval Dashboard` and select the **My Approvals** tab.
    |   |   |
    |---|---|

>     | ![Navigation1 Dashboard](../../media/development/approval-workflows/nav-approval-dashboard1.png) | ![Navigation2 Dashboard](../../media/development/approval-workflows/nav-approval-dashboard2.png) |

Reviewer: This isn't rendering correctly either. If you want to be able to put an ordered list or other multi-line content inside a table, you may have to use raw HTML instead of Markdown.

-d '{"comment": "Deny reason"}' \
http://nautobot/api/extras/approval-workflow-stages/$APPROVAL_WORKFLOW_STAGE_ID/deny
```

#### List Pending/Done Approvals

Retrieves a list of approval workflow stages filtered by their status relative to the current user using the `pending_my_approvals` query parameter on the standard list endpoint:

- `?pending_my_approvals=true` — Returns stages pending approval by the current user.

```no-highlight
curl -X GET \
-H "Authorization: Token $TOKEN" \
-H "Accept: application/json; version=1.3; indent=4" \
http://nautobot/api/extras/approval-workflow-stages/?pending_my_approvals=true
```

- `?pending_my_approvals=false` — Returns stages the current user has already approved/denied.

```no-highlight
curl -X GET \
-H "Authorization: Token $TOKEN" \
-H "Accept: application/json; version=1.3; indent=4" \
http://nautobot/api/extras/approval-workflow-stages/?pending_my_approvals=false
```

If the parameter is omitted, all stages are returned regardless of approval status.

## Permissions by Persona

### Object Operator

Responsible for creating, updating, or running objects, which may trigger an approval workflow. Cannot approve workflows.

> **Required permissions:**
> - `extras.view_approvalworkflow` - to view created approval workflows after creating, updating, or running an object.
> - `extras.view_approvalworkflowstage` - to view the **My Requests** tab on the Dashboard.
> - `extras.view_approvalworkflowstageresponse` - to view responses for each stage.

Reviewer: Not rendering correctly, I think this will fix it:

```suggestion
**Required permissions:**

- `extras.view_approvalworkflow` - to view created approval workflows after creating, updating, or running an object.
- `extras.view_approvalworkflowstage` - to view the **My Requests** tab on the Dashboard.
- `extras.view_approvalworkflowstageresponse` - to view responses for each stage.
```

- Additional permissions:
    - `extras.view/add/change/delete_approvalworkflowdefinition`
    - `extras.view/add/change/delete_approvalworkflowstagedefinition`

### Approver Groups

Approval actions are controlled not only by permissions but also by **approver group membership** defined in each workflow definition:

- Each **Approval Workflow Stage Definition** specifies an **Approver Group** (user group authorized to act on that stage, how to create user group via the admin UI check [here](users/groups.md)).
- A user must belong to the stage’s approver group to:
    - See the stage in the **My Approvals** tab.
    - Perform approval or denial actions.
- If the user has the required permissions but is **not** a member of the approver group, the stage will not appear in their dashboard and no actions will be available.

## Upgrade Considerations

### From Nautobot 2.x

If upgrading from Nautobot 2.x, the management command `check_job_approval_status` is available to identify Jobs and Scheduled Jobs that have `approval_required=True`.

- Running the command doesn't approve/run/deny jobs, it just identifies the ones that need such action to be performed as a separate step.

> - Before upgrading, approval workflows aren't a thing, defining those needs to be done after the upgrade.

Reviewer: ```suggestion
- After running this command to identify the impacted Jobs, and completing the upgrade to Nautobot 3.x, you'll want to define appropriate approval workflows to apply to those jobs.
```

> **Example**: User running command `check_job_approval_status` get output like

Reviewer 2: ```suggestion
**Example**: Running the command `check_job_approval_status` will provide similar output to:
```

```no-highlight
Following jobs still have `approval_required=True`.
These jobs will no longer trigger approval automatically.
After upgrading to Nautobot 3.x, you should add an approval workflow definition(s) covering these jobs.
Refer to the documentation: https://docs.nautobot.com/projects/core/en/next/user-guide/platform-functionality/approval-workflow/
Affected jobs (Names):
    - ExampleDryRunJob
    - Example Job of Everything
    - Export Object List
```

- User can create an `ApprovalWorkflowDefinition` using constraints such as `{"job_model__name__in": ["ExampleDryRunJob", "Example Job of Everything", "Export Object List"]}`. See [Create an Approval Workflow Definition with stages](#create-an-approval-workflow-definition-with-stages)
- However, constraints today are limited to what can be expressed as a single dictionary of field lookups. More complex logic (e.g., `name=A OR (name=B AND status=active)`) is not possible yet.

> - Support for combining multiple conditions with **Q objects** and richer filtering logic is planned for **Nautobot 3.1**.

Reviewer: Let's not pre-commit ourselves here. :)

```suggestion
```

        GenericForeignKey object_under_review FK
        ContentType object_under_review_content_type FK
        uuid object_under_review_object_id
        choices current_state
        datetime decision_date
        User user FK
        string user_name
    }
    "extras.ApprovalWorkflowStage" {
        ApprovalWorkflow approval_workflow FK
        ApprovalWorkflowStageDefinition approval_workflow_stage_definition FK
        choices state
        datetime decision_date
    }
    "extras.ApprovalWorkflowStageResponse" {
        ApprovalWorkflowStage approval_workflow_stage FK
        User user FK
        string comments
        choices state
    }

    "contenttypes.ContentType"[ContentType] {}
    "users.User"[User] {}
    "users.Group"[Group] {}
    "models.Model"[Model] {}

    "extras.ApprovalWorkflowDefinition" }o--|| "contenttypes.ContentType": "applies to model"
    "extras.ApprovalWorkflowDefinition" ||--o{ "extras.ApprovalWorkflowStageDefinition": "defines stages"
    "extras.ApprovalWorkflowStageDefinition" }o--|| "users.Group": "approver group"
    "extras.ApprovalWorkflowDefinition" ||--o{ "extras.ApprovalWorkflow": "creates instances"

>     "extras.ApprovalWorkflow" }o--|| "models.Model": "object under review"

Author: No, we don’t literally have a Django model class named `Model` in Nautobot’s code. In the ERD, "models.Model" is just a generic placeholder representing any Django model that can be the target of the `object_under_review` GenericForeignKey. It’s not an actual database table in Nautobot.

Django’s own ORM has a base class `django.db.models.Model` that all models inherit from. That’s what the diagram is hinting at.

>     "extras.ApprovalWorkflow" }o--|| "models.Model": "object under review"

Reviewer 2: @smk4664 any thoughts on how to represent a GFK?
